remove category label requirement from PRs

cburroughs opened the pull request:
The category label used to be used as part of release notes generation, but now that we a spiffy new releases notes process <https://github.com/pantsbuild/pants/discussions/20888> and they are no longer used for "major" releases.

They were still part of the GitHub patch releases notes but:
 * Bugfix releases ought to be all bug fixes already
 * It is a point of friction for contributors; every first time contributor is likely to get a red X and it is hard for reviewers to tell at a glance from the PR list which failures are real.

So on net, the benefit is no longer worth the friction.

This diff renders them optional by:
 * Removing the Action check.
 * Defaulting to not using them for headings in the GitHub release notes
 * Removing from docs.
  
 After using this for a bit we can see if it feels right and if omitting `internal` changes from the GitHub release description is worth the vestigial support.

benjyw reviewed (commented):
"They are still part of the github patch releases notes" - and `git grep '"category:'` shows that these are still expected by certain scripts. Presumably we should remove all that first, or in this PR?

cburroughs commented:
> "They are still part of the github patch releases notes" - and `git grep '"category:'` shows that these are still expected by certain scripts. Presumably we should remove all that first, or in this PR?

The use in `changelog.py` was optional due to the default `None` passthru, but I've gone ahead and made them explicitly optional and tidied up the output.

benjyw reviewed (approved):
Seems fine, but why not nuke all uses of the labels entirely?

cburroughs commented:
> why not nuke all uses of the labels entirely?

I thought that was your suggestion from the prior review comments.  That is leaving it as mishmash like:

```
### Bug Fixes

 * Fixed a bug (arbitrarily used label)
 
 ## Performance
 
  * Rewrote everything in Rust (arbitrarily used a label)
  
  ## Uncategorized
  
   * Removed fruit fly from vacuum tube (did not use label)
   * Replaced used of bubble sort (did not use label)
```

Would look kind of awkward.

cburroughs commented:
They are used today as the Markdown headings in the *GitHub release* description and to exclude `internal` commits from the same.  An example:  https://github.com/pantsbuild/pants/releases/tag/release_2.24.0.dev1

For the `x.y.0` releases we point towards the human written release notes: https://github.com/pantsbuild/pants/releases/tag/release_2.22.0  But the GitHub descriptions are still what we have for patch releases.  (There is a UX gap today where the final patch release doesn't list the rc changes, but I think that is orthogonal to the categories #20173)

I'm arguing the headings are not worth the new contributor toil and we can stop generating them (have one flat list).

benjyw commented:
> 
> I'm arguing the headings are not worth the new contributor toil and we can stop generating them (have one flat list).

Seems right to me. So now I'm back to my original question of why not remove that entirely from src/python/pants_release/changelog.py, rather than have that functionality retained, but optional? Does that make sense, or am I missing something.

cburroughs commented:
>  rather than have that functionality retained, but optional? 

Gotcha.  I started to rip it out entirely and then realized that in addition to adding the headings we filter out `category:internal`.  So something like https://github.com/pantsbuild/pants/releases/tag/release_2.24.0.dev1 looks different with all of the PyO3 commits included.  That was (somewhat) bigger presentation changes than "just remove the headers" than I had intended.

The new logic of `category:internal` OR `release-notes:not-required` requires almost all of the same scaffolding to hit the GitHub API so leaving the prior logic as optional felt like the smallest "change one thing at a time".

I think "What makes the GitHub release description useful to users and what should they look like?" is a fair question, but not exactly the one I had intended to answer originally.

benjyw commented:
Gotcha. I think just supporting `release-notes:not-required` should be enough eventually.
